### ISSUE TITLE
Allow selection of energy axis slices in MapDataset.to_image?

### ISSUE THREAD

**Issue by lmohrmann** [gammapy/gammapy#2701]
This is mostly a convenience feature, but should not be very hard to implement.

**Current behaviour**
When calling `MapDataset.to_image()`, the maps are summed over all energy bins.

**Desired behaviour**
It would be convenient if one could pass a slice to be applied to the energy axis, thus selecting which energy bins should be summed over.

**Comment by adonath** [comment]
Technically this is currently handled with the `.mask_safe` attribute, which is applied before computing the image dataset.

We can certainly extend the current behaviour to also include the `mask_fit`, so that an energy range can be chosen like so:
```
dataset.mask_fit = dataset.counts.geom.energy_mask(emin=, emax=)
dataset_image = dataset.to_image() 
```
This would already be an improvement.

If we implement slicing in addition I'd prefer to mirror the Map API and implement a `MapDataset.slice_by_idx()` or `.slice_by_coord()` method, which is much more flexible then adding this functionality in a single place only...

**Comment by lmohrmann** [comment]
@adonath Thanks for the explanation! I hadn't realized that it's apparently possible to achieve this via the `safe_mask`.

The idea of adding e.g. `slice_by_idx` for datasets sounds good, that might be convenient. I will not have time to implement something like this, so I'll leave it to you to either close this issue or leave it open in case you would like to see this feature implemented. Either is fine with me.

**Comment by registerrier** [comment]
I agree that adding the `slice` method is useful since a user might need a slice of the `Dataset`, which is a more general use case than taking the image of the slice. This would have to be implemented on `SpectrumDataset` as well.

A remark though, since datasets are always binned in energy a `slice_by_coord` method might be unintuitive.

**Comment by adonath** [comment]
@registerrier I still agree it's useful (e.g. to speed up flux point estimation), but I also realised it's not as simple, because we have to deal somehow with the reconstructed and true energy at the same time. So do we slice in reco energy? Where do we cut the true energy then? I have no clear proposal here yet...

**Comment by registerrier** [comment]
You're right. This probably means one has to keep all true energy bins anyways. Masking seems a simpler idea, then. We should document more precisely how to use this.

**Comment by adonath** [comment]
In #2966 we implemented the possibility to slice the energy axis of a `MapDataset` using `MapDataset.slice_by_idx()`. The behaviour nows is that the slicing only applies to the given axis. In addition I also added `Datasets.slice_energy()`, which leaves the truth energy axis untouched for now...as a convenience we could add `Dataset.slice_energy()` as well and maybe add an option to give a PDF threshold, which is used to slice the true energy axis accordingly...

**Comment by adonath** [comment]
@registerrier We could add a method `MapDataset.slice_energy(emin, emax, edisp_pdf_threshold=)` which does the energy slicing and optionally slices the true energy axis as well, bases on a certain pdf_threshold for the edisp. This might be interesting to speed up computations of flux points...

**Comment by registerrier** [comment]
Yes it could be done. I don't know if `pdf_threshold` is sufficient/stable enough or if one should impose a fraction of the total pdf included (e.g. 99%).

Note that, this might done also at a lower level to speed up all calculations. One could replace the `.dot` product in `apply_edisp`.

**Comment by adonath** [comment]
@registerrier I think the lowest level we could go is the `MapEvaluator`, because the edisp is position dependent. On the same go where we cutout the exposure map per model component we could slice the true energy axis as well. However I'm not sure how much of an effect this has on performance...
